File: frontend/onied/src/components/blocks/video/providers/rutubeVideoProvider.ts
import VideoProvider from "./videoProvider";

/**
 * Rutube Video Provider.
 *
 * @class RutubeVideoProvider
 * @extends {VideoProvider}
 */
export default class RutubeVideoProvider extends VideoProvider {
  regex =
    /^((?:https?:)?\/\/)?(rutube\.ru)(\/video)(?<videoId>\/[\w\d]+)?(\/[\S]+)?$/;

  getLink(href: string) {
    const matches = href.match(this.regex);

    return Promise.resolve(
      `https://rutube.ru/play/embed/${matches?.groups?.videoId}`
    );
  }
}

File: mobile/onied/lib/views/course/components/blocks/videoBlock/video/providers/rutube_video_provider.dart
import 'video_provider.dart';

class RutubeVideoProvider extends VideoProvider {
  @override
  RegExp get regex => RegExp(
    r'^((?:https?:)?\/\/)?(rutube\.ru)(\/video)(?<videoId>\/[\w\d]+)?(\/[\S]+)?$',
  );

  @override
  String getLink(String href) {
    final matches = regex.firstMatch(href);
    if (matches == null || matches.namedGroup('videoId') == null) {
      throw Exception('Невалидный URL Rutube');
    }

    final videoId = matches.namedGroup('videoId')!;
    return 'https://rutube.ru/play/embed$videoId';
  }
}
